### DISCUSSION TITLE
blog/structured-logging-with-serilog-in-aspnet-core/

### DISCUSSION THREAD

**Discussion by giscus** [codewithmukesh/discussions#5] [General]
# blog/structured-logging-with-serilog-in-aspnet-core/

When building a new ASP.NET Core project, one should prioritize setting up logging to ensure robust monitoring and debugging capabilities right from the start. Serilog is the most popular logging library for ASP.NET Core Applications. In this article, we will learn everything you need to know to master Structured Logging in your ASP.NET Core Application using Serilog.

https://codewithmukesh.com/blog/structured-logging-with-serilog-in-aspnet-core/?utm_source=codewithmukesh.beehiiv.com&utm_medium=newsletter&utm_campaign=structured-logging-with-serilog-in-asp-net-core

**Comment by Gooseman** [comment]
You correclty point out that using asynchronous logging is a best practice, but you haven't given an information on how to tell Serilog to log asynchronously.  Is there a way to enable asynchronous logging without hard-coding the log setup?  I.e. is there a way to put it into the config file rather than directly into the code via WriteTo.Async?

**Reply by iammukeshm** [reply]
Yes, there is. Will include it in the article. However, here is the config needed,

```
{
  "Serilog": {
    "WriteTo": [{
      "Name": "Async",
      "Args": {
        "configure": [{
          "Name": "Console"
        }]
      }
    }]
  }
}
```
Refer: https://github.com/serilog/serilog-sinks-async

**Reply by nicky2983** [reply]
How to enable asynchronous when writing Seq? Or is it already asynchronous? Thank you

**Comment by nicky2983** [comment]
Can I log api request / response json? I have tried to use Enrich "FromLogContext", but seems not work

**Reply by hmstrand** [reply]
Check so you don't have a higher log level configure which will override the request log from serilog. For example: 

```
configuration
    .MinimumLevel.Error()
```

Then you need to override this for Serilog.AspNetCore.RequestLoggingMiddleware.

```
.MinimumLevel.Override("Serilog.AspNetCore.RequestLoggingMiddleware", LogEventLevel.Information)
```

**Comment by N-Nagaraj** [comment]
How to configure the tenant specific logs store local development
 "WriteTo": [
      {
        "Name": "Async",
        "Args": {
          "configure": [
            {
              "Name": "Map",
              "Args": {
                "keyPropertyName": "TenantId",
                "default": "logs/default/log-.txt",
                "createSink": {
                  "Name": "File",
                  "Args": {
                    "path": "logs/{TenantId}/log-.txt",
                    "rollingInterval": "Day",
                    "retainedFileCountLimit": 30,
                    "fileSizeLimitBytes": 10485760,
                    "formatter": "Serilog.Formatting.Json.JsonFormatter, Serilog"
                  }
                }
              }
            }
          ]
        }
      }
    ]

I used above it's not working

**Comment by shady78** [comment]
thanks

**Comment by harinadh8n** [comment]
How to do loggin in database

**Comment by mohamedHodaib** [comment]
Here is the complete and best practice logging configuration:No, the provided Serilog configuration is **not complete** and **doesn't fully represent best practices**.

It's a good start, but it lacks several key elements commonly used in robust, production-ready logging setups.

-----

## Missing Key Elements

The configuration you provided is missing several important components often considered best practice for a complete logging setup:

### 1\. Output Sink Details (The File Sink) 📄

The `File` sink configuration is incomplete. It's missing the crucial **output path** and **file name template**.

  * **Missing `path` or `pathFormat`**: You need to tell Serilog *where* to write the log files.
  * **Missing formatter**: While not strictly required, specifying an **output template** or using a structured format like **JSON** is a best practice, especially if logs are to be indexed by a log management system (LMS).

### 2\. Output Sinks (External Storage) ☁️

Best practice dictates logging to more than just a local file, especially for production environments. Logging to external sinks is essential for centralized analysis and alerting.

  * **Console Sink**: Essential for development and containerized environments (like Docker/Kubernetes).
  * **External Sink**: A sink like **Seq**, **Elasticsearch (via Serilog.Sinks.Elasticsearch)**, **Application Insights**, or a cloud service sink (e.g., **Azure Log Analytics**, **AWS CloudWatch**) is standard for production-grade, centralized logging.

### 3\. Enrichment 🏷️

Enrichers add valuable context to every log event. Common and essential enrichers are missing:

  * **Machine Name (`WithMachineName`)**: Identifies the server where the log originated.
  * **Thread Id/Process Id (`WithThreadId`, `WithProcessId`)**: Helpful for tracing concurrent operations.
  * **Source Context (`FromLogContext`)**: Essential for capturing the fully-qualified class/namespace name, making it easier to filter and diagnose issues.

-----

## Recommended Best Practice Configuration Example

A more complete, best-practice configuration often looks something like this (using a common set of sinks and enrichers):

```json
{
  "Serilog": {
    "Using": [ "Serilog.Sinks.Console", "Serilog.Sinks.File", "Serilog.Enrichers.Environment", "Serilog.Enrichers.Process" ], // Add required packages
    "MinimumLevel": {
      "Default": "Information",
      "Override": {
        "Microsoft": "Warning",
        "Microsoft.AspNetCore.Hosting.Diagnostics": "Error",
        "System": "Warning" // Good to quiet down System namespace too
      }
    },
    "Enrich": [ "FromLogContext", "WithMachineName", "WithProcessId", "WithThreadId" ], // Add enrichers
    "WriteTo": [
      // 1. Console Sink (For Development/Containers)
      {
        "Name": "Console",
        "Args": {
          "outputTemplate": "[{Timestamp:HH:mm:ss} {Level:u3}] {Message:lj}{NewLine}{Exception}"
        }
      },
      // 2. File Sink (For Local Persistence/Backup)
      {
        "Name": "File",
        "Args": {
          "path": "C:\\logs\\myapp\\log-.txt", // IMPORTANT: Define the path!
          "rollingInterval": "Day",
          "rollOnFileSizeLimit": true,
          "fileSizeLimitBytes": 10485760, // 10 MB limit
          "formatter": "Serilog.Formatting.Compact.CompactJsonFormatter, Serilog.Formatting.Compact" // Structured JSON format
        }
      }
      // 3. External Sink (For Centralized Logging - e.g., Seq, Elasticsearch, Log Analytics)
      /*
      {
        "Name": "Seq",
        "Args": {
          "serverUrl": "http://localhost:5341"
        }
      }
      */
    ]
  }
}
```

**Comment by YashSmithh** [comment]
To implement Serilog in ASP.NET Core, install the Serilog NuGet packages, configure sinks (like console, file, or Seq), and integrate it in Program.cs using UseSerilog(). You can enrich logs with contextual information such as request IDs, user data, or environment details for better observability.

**Comment by fahim5466** [comment]
This was a good read. Thanks!
